pkg/kv/kvserver/rangefeed/scheduler.go

```diff
@@ -200,76 +253,86 @@ func (s *Scheduler) Register(f Callback) (int64, error) {
 // that processor actually handled stopped event it may either be pending or
 // processed.
 func (s *Scheduler) Enqueue(id int64, evt processorEventType) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	if _, ok := s.mu.procs[id]; !ok {
+	shard := s.shards[shardIndex(id, len(s.shards))]
+
+	shard.Lock()
+	defer shard.Unlock()
+	if _, ok := shard.procs[id]; !ok {
 		return
 	}
-	newWork := s.enqueueInternalLocked(id, evt)
+	newWork := shard.enqueueLocked(id, evt)
 	if newWork {
 		// Wake up potential waiting worker.
 		// We are allowed to do this under cond lock.
-		s.cond.Signal()
+		shard.cond.Signal()
 	}
 }
 
-func (s *Scheduler) enqueueInternalLocked(id int64, evt processorEventType) bool {
-	pending := s.mu.status[id]
+// enqueueLocked enqueues a single event for a given processor in this shard.
+func (ss *schedulerShard) enqueueLocked(id int64, evt processorEventType) bool {
+	pending := ss.status[id]
 	if pending&Stopped != 0 {
 		return false
 	}
 	if pending == 0 {
 		// Enqueue if processor was idle.
-		s.mu.queue.pushBack(id)
+		ss.queue.pushBack(id)
 	}
 	update := pending | evt | Queued
 	if update != pending {
 		// Only update if event actually changed.
-		s.mu.status[id] = update
+		ss.status[id] = update
 	}
 	return pending == 0
 }
 
-// EnqueueAll enqueues event for all existing non-stopped id's. Enqueueing is
-// done in chunks to avoid holding lock for too long and interfering with other
-// enqueue operations.
-//
-// If id is not known or already stopped it is ignored.
-func (s *Scheduler) EnqueueAll(ids []int64, evt processorEventType) {
-	scheduleChunk := func(chunk []int64) int {
-		s.mu.Lock()
-		defer s.mu.Unlock()
-		wake := 0
-		for _, id := range chunk {
-			if _, ok := s.mu.procs[id]; ok {
-				if newWork := s.enqueueInternalLocked(id, evt); newWork {
-					wake++
-				}
-			}
-		}
-		return wake
+// enqueueN enqueues an event for multiple processors on this shard. The
+// caller must not hold the shard lock.
+func (ss *schedulerShard) enqueueN(ids []int64, evt processorEventType) int {
+	// Avoid locking for 0 new processors.
+	if len(ids) == 0 {
```

Comment:
We can avoid calling this for empty shards if we make the check before the call. Saves us unnecessary call.

Reply:
Yup, done.